### DISCUSSION TITLE
React Hooks version of react-collapsible

### DISCUSSION THREAD

**Discussion by glennflanagan** [glennflanagan/react-collapsible#234] [Announcements]
@iMoses has created a really nice-looking Gist of the way React Collapsible could move forward using React Hooks (https://github.com/glennflanagan/react-collapsible/issues/228)

If we can we would want this to be backwards compatible - or at least still offer a version of React Collapsible that doesn't require React Hooks.

This discussion is open to all who would like to offer advice/requirements and/or like to help out move this forward.

**Comment by gioragutt** [comment]
Basically if I understand correctly the book contains all the logic contained by `Collapsible`. The refactor could be that `Collapsible` would use the hook under-the-hood, and so you'll have both the Hook and the Component versions.

**Comment by iMoses** [comment]
I think it shouldn't be a problem to create the exact same interface and output using a hook under the hood, question is if we want to create an exact replica which is 100% backward compatible, maybe polish the interface a little or potentially do/offer both.

I'll work on a compatible component and will provide an example soon...

**Reply by iMoses** [reply]
Here`s a backward compatible [example](https://gist.github.com/iMoses/24b9dd6e13a26145a919127d38f8c088#file-backwards-compatible-example-js)

The only noticable difference is that inside `handleTriggerClick` I reversed the order of these:
```
    if (this.props.triggerDisabled || this.state.inTransition) {
      return;
    }

    event.preventDefault();
```
so that even when disabled we'll `preventDefault` to keep consistency (which at least for us caused a bug before).
other than that they should be exactly the same.

These changes could also be found here: https://github.com/iMoses/react-collapsible/tree/imoses/v3-rc

**Reply by glennflanagan** [reply]
Wow. This is incredible work. I’ll try and review it fully asap. 

I think we should put this out as v3 to try and mitigate any unforeseen bugs by doing this. 

Great work!!

**Reply by iMoses** [reply]
I fixed the broken tests and have found another difference between the two implementations.
In the current implementation if the `open` property is changed it will automatically change the state counterpart, that's because `componentDidUpdate` will allow you to change the `state` based on the recieved props before the render method is called, but when using function components this takes another tick to apply to change since updating the state is done asynchronous when using hooks.

So if you pull from my fork and run the tests this will be the only one to fail. 
I didn't wanna change it before consulting with you on the approach.

**Edit**
Maybe I was wrong and the fix is simply to update the component using `collapsible.update()`
Please validate me on that one, I'm not that strong with Enzyme.
I've pushed this fix that makes all tests pass.

**Reply by karltaylor** [reply]
Hey @iMoses might be a big ask but is there anyway we can build it in typescript?

**Reply by karltaylor** [reply]
Also I think I tried to add this to a previous v3 to change the way some of the class logic worked:

https://github.com/glennflanagan/react-collapsible/issues/87

**Reply by iMoses** [reply]
I don't mind adding TS, but if that's the case I'll probably upgrade all the infrastructure.

Regarding #87 - as I previously said I think that v3 should export 3 entities. The hook, a 100% backwards compatible component utilizing the hook, and a new version of the component which introduces a cleaner interface which will not be compatible (although migrating would be easy)

and I think that regarding the entry points it should be that `react-collapsible` would provide the new version while `react-collapsible/legacy` (or something of the sort) would provide the compatible version so that the migration between v2 and v3 could simply be search & replacing the import if they don't wish to change anything right away.

WDYT?

I'll come back with a code example once I have the time :)

**Reply by karltaylor** [reply]
I'm not sure it _needs_ to be backwards compatible, we can simply release a new major version `3.0.0`, I think doing a `/legacy` thing might get a little convoluted for the contributors and for community using it, don't you think?

**Reply by iMoses** [reply]
@karltaylor sorry, I missed the last message
I tend to agree with you, I think a "legacy" version can be "exampled" as part of the documentation, outside of the package.


@glennflanagan note that a typescript version of the code is sitting idle for a couple of months now :)
I'm still in favor of publishing it and am willing to put in the time
I only require your consent and review

https://github.com/iMoses/react-collapsible/blob/imoses/v3-rc/src/useCollapsible.ts

**Reply by gioragutt** [reply]
Hi all 😀 any update on this? Would love to use the hook version in my codebase!
